### PR TITLE
[Traceable FSDP2] Change from register_multi_grad_hook to per-tensor backward hook

### PR DESCRIPTION
**Pull request by yf225**
As discussed with Andrew before, under compile we will register per-tensor backward hook instead of multi-grad hook, because it's difficult for Dynamo to support `register_multi_grad_hook` (or anything `.grad_fn` related). We expect both to have the same underlying behavior, ~~and we will add integration test (in subsequent PR) to show that compile and eager has same numerics.~~

As discussed below, we will change eager path to use per-tensor backward hook as well.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

**Comment by awgu** [comment]
Thanks @yf225 for the PR!

Can we try to do the following instead:
1. Change `FSDPState._pre_backward` to return early if already in `TRAINING_STATE.PRE_BACKWARD`
2. Use `register_hook` (not `register_multi_grad_hook`) for eager path

I would prefer to avoid the divergence, and if we are betting on the two being the same, then we might as well make them the same.

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 1]
```diff
@@ -235,7 +244,13 @@ def _register_pre_backward_hook(self, output: Any) -> Any:
             t for t in flat_outputs if (torch.is_tensor(t) and t.requires_grad)
         )
         if tensors:
-            register_multi_grad_hook(tensors, self._pre_backward, mode="any")
+            # Invariant: if Compiled Autograd is used, regardless of whether forward is eager mode,
+            # we must use the Traceable FSDP codepath.
+            if not torch._dynamo.compiled_autograd.compiled_autograd_enabled:
```
Comment (fegin) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:249]: Is this flag going to be True even if users only wrap the `loss.sum().backward()` with CompiledAutograd?

Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:249]: For Traceable FSDP case, from my experiment it seems we need to wrap the forward with CompiledAutograd as well.

I feel that it might also be a good convention to use going forward (i.e. "if you want to use PT2-D, please wrap both forward and backward with CompiledAutograd").

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 2]
```diff
@@ -235,7 +237,8 @@ def _register_pre_backward_hook(self, output: Any) -> Any:
             t for t in flat_outputs if (torch.is_tensor(t) and t.requires_grad)
         )
         if tensors:
-            register_multi_grad_hook(tensors, self._pre_backward, mode="any")
+            for tensor in tensors:
+                handle = tensor.register_hook(self._pre_backward)
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:241]: nit:
```suggestion
                tensor.register_hook(self._pre_backward)
```

**Comment by awgu** [comment]
@yf225 I think the issue is that `FSDPState` does not have a `POST_BACKWARD` training state since "post-backward" is specific to `FSDPParamGroup`, which runs post-backward logic.

I think it might work if we instead change the early return to `FSDPParamGroup.pre_backward`.

**Comment by yf225** [comment]
> @yf225 I think the issue is that `FSDPState` does not have a `POST_BACKWARD` training state since "post-backward" is specific to `FSDPParamGroup`, which runs post-backward logic.
> 
> I think it might work if we instead change the early return to `FSDPParamGroup.pre_backward`.

Thanks @awgu! Yes I believe this works from my local testing, gonna try CI again

**Comment by yf225** [comment]
CI error seems unrelated, gonna rebase

**Comment by yf225** [comment]
cc. @awgu looks like all tests passed! :)

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 3]
```diff
@@ -201,11 +200,17 @@ def _post_forward(self, module: nn.Module, input: Any, output: Any) -> Any:
...
+        if (
+            self._fsdp_param_group
+            and self._fsdp_param_group._training_state == TrainingState.PRE_BACKWARD
+        ):
+            return grad
         self._training_state = TrainingState.PRE_BACKWARD
         self._register_root_post_backward_final_callback()
         if self._fsdp_param_group:
-            self._fsdp_param_group.pre_backward(*unused)
+            self._fsdp_param_group.pre_backward()
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:212]: nit: What did you think of making the early return in `FSDPParamGroup.pre_backward`?

It is less intuitive that `FSDPState._pre_backward` should be checking its `self._fsdp_param_group._training_state` to affect behavior. In particular, I am suggesting deleting 204-208, and moving the check to `FSDPParamGroup.pre_backward` to return early if already `self._training_state == TrainingState.PRE_BACKWARD`.

**Review by awgu — APPROVED** [review]
Stamp to unblock but one thing worth considering
